packages/smart-contracts/src/contracts/EthereumSingleRequestProxy.sol

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import './interfaces/EthereumFeeProxy.sol';
+
+/**
+ * @title EthereumSingleRequestProxy
+ * @notice This contract is used to send a single request to a payee with a fee sent to a third address
+ */
+contract EthereumSingleRequestProxy {
+  address public payee;
+  bytes public paymentReference;
+  address public feeAddress;
+  uint256 public feeAmount;
+  IEthereumFeeProxy public ethereumFeeProxy;
+
+  address private originalSender;
+
+  // Reentrancy guard
+  uint256 private constant _NOT_ENTERED = 1;
+  uint256 private constant _ENTERED = 2;
+  uint256 private _status;
+
+  constructor(
+    address _payee,
+    bytes memory _paymentReference,
+    address _ethereumFeeProxy,
+    address _feeAddress,
+    uint256 _feeAmount
+  ) {
+    payee = _payee;
+    paymentReference = _paymentReference;
+    feeAddress = _feeAddress;
+    feeAmount = _feeAmount;
+    ethereumFeeProxy = IEthereumFeeProxy(_ethereumFeeProxy);
+    _status = _NOT_ENTERED;
+  }
+
+  modifier nonReentrant() {
+    if (msg.sender != address(ethereumFeeProxy)) {
+      // On the first call to nonReentrant, _status will be _NOT_ENTERED
+      require(_status != _ENTERED, 'ReentrancyGuard: reentrant call');
+      // Any calls to nonReentrant after this point will fail
+      _status = _ENTERED;
+    }
+    _;
+    if (msg.sender != address(ethereumFeeProxy)) {
+      // By storing the original value once again, a refund is triggered
+      _status = _NOT_ENTERED;
+    }
+  }
+
+  receive() external payable nonReentrant {
+    if (msg.sender == address(ethereumFeeProxy)) {
+      // Funds are being sent back from EthereumFeeProxy
+      require(originalSender != address(0), 'No original sender stored');
+
+      // Forward the funds to the original sender
+      (bool forwardSuccess, ) = payable(originalSender).call{value: msg.value}('');
+      require(forwardSuccess, 'Forwarding to original sender failed');
+
+      // Clear the stored original sender
+      originalSender = address(0);
+    } else {
+      require(originalSender == address(0), 'Another request is in progress');
+
+      originalSender = msg.sender;
+
+      bytes memory data = abi.encodeWithSignature(
+        'transferWithReferenceAndFee(address,bytes,uint256,address)',
+        payable(payee),
+        paymentReference,
+        feeAmount,
```

Comment:
In both `ERC20SingleRequestProxy` and `EthereumSingleRequestProxy` the fees amount are paid every time the `receive` method is called. 
It won't be possible for a user to pay his requests with multiple payment as he'll need to pay for the fees several times. Is it the intended behavior ?

Reply:
Yes, that is the intended behavior. The same thing happens when a request is created with `feeAddress` and `feeAmount` using the SDK; the fee is paid every time there is a payment.

This is a test request we did to see the behavior of the partial payments with fees , you can see the service fee being paid every time:
https://scan.request.network/request/0118bc428d569e48a62a64625bbd836dfc00c935caae8a201daf308438fce3ddd9

Reply:
The fees aren't paid necessarely every time there is a payment. For example if I have a request where I need to pay 100$ as well as 2$ of fees, I can:
- Make on payment of 50$ and 2$ fees.
- Another payment of 50$ and 0$ of fees
We'll detect both of these payments and the request will be considered paid, but with the current implementation of the contract you'd have to pay twice the 2$ of fees if you want to make partial payments.
Maybe the `feeAmount` should be updated after each payment ?

I don't understand the link you shared, the payment are not partial, each of them account for the full amount of the request (0.003 ETH)